File: app/src/main/java/com/ad6f/bowling/services/sensors/SensorManagement.java
package com.ad6f.bowling.services.sensors;

import android.hardware.Sensor;
import android.hardware.SensorManager;

public class SensorManagement {
    private boolean isRegister = false;

    private SensorManager sensorManager;
    
    private SensorCalculator linearCalculator;

    private SensorCalculator rotationCalculator;

    private SensorCalculator gravityCalculator;
    
    public void unregisterListeners() {
        if(isRegister) {
            linearCalculator.unregisterListener(sensorManager);
            rotationCalculator.unregisterListener(sensorManager);
            gravityCalculator.unregisterListener(sensorManager);
            isRegister = false;
        }
    }

    public void registerListeners() {
        if(!isRegister) {
            linearCalculator.registerListener(sensorManager);
            rotationCalculator.registerListener(sensorManager);
            gravityCalculator.registerListener(sensorManager);
            isRegister = true;
        }
    }

    public void start() {
        linearCalculator.start();
        rotationCalculator.start();
        gravityCalculator.start();
    }

    public void close() {
        linearCalculator.close();
        rotationCalculator.close();
        gravityCalculator.close();
    }

    public float getForce() {
        return linearCalculator.getBestPercent();
    }

    public float getTilt() {
        return gravityCalculator.getLastPercent();
    }

    public float getRotation() {
        return rotationCalculator.getBestPercent();
    }

    public SensorManagement(SensorManager sensorManager, SensorAction sendData) {
        this.sensorManager = sensorManager;
        
        this.linearCalculator = new SensorCalculator(
            sensorManager, 
            Sensor.TYPE_LINEAR_ACCELERATION,
            Coordinate.Y,
            sendData
        );

        this.rotationCalculator = new SensorCalculator(
            sensorManager,
            Sensor.TYPE_ROTATION_VECTOR,
            Coordinate.Y,
            null
        );

        this.gravityCalculator = new SensorCalculator(
            sensorManager,
            Sensor.TYPE_ROTATION_VECTOR,
            Coordinate.X,
            null
        );

        this.registerListeners();
    }
}

File: app/src/main/java/com/ad6f/bowling/services/sensors/SensorManagement.kt
package com.ad6f.bowling.services.sensors

import android.hardware.Sensor
import android.hardware.SensorManager

class SensorManagement(private val sensorManager: SensorManager, sendData: () -> Unit) {
    private var isRegister = false

    private val linearCalculator = SensorCalculator(
        sensor = sensorManager.getDefaultSensor(Sensor.TYPE_LINEAR_ACCELERATION)!!,
        coordinate = Coordinate.Y,
        sendData
    )

    private val rotationCalculator = SensorCalculator(
        sensor = sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)!!,
        coordinate = Coordinate.Y
    )

    private val gravityCalculator = SensorCalculator(
        sensor = sensorManager.getDefaultSensor(Sensor.TYPE_GRAVITY)!!,
        coordinate = Coordinate.X
    )

    fun start() {
        if (isRegister) return

        linearCalculator.registerListener(sensorManager)
        rotationCalculator.registerListener(sensorManager)
        gravityCalculator.registerListener(sensorManager)
        isRegister = true
    }

    fun close() {
        if (!isRegister) return

        linearCalculator.unregisterListener(sensorManager)
        rotationCalculator.unregisterListener(sensorManager)
        gravityCalculator.unregisterListener(sensorManager)
        isRegister = false
    }

    fun getJsonData(): String {
        return "{\"rotation\":${rotationCalculator.getBestPercent()}, \"force\":${linearCalculator.getBestPercent()}, \"tilt\":${gravityCalculator.getLastPercent()}}"
    }
}
